lib/queuePopulator/IngestionProducer.js

```diff
@@ -185,95 +256,142 @@ class IngestionProducer {
+     * @param {string} [state.keyMarker] - NextKeyMarker
      * @param {function} done - callback function
-     * @return {object} list of buckets and list of objects for each bucket
+     * @return {object} list of objects for each bucket, including a duplicate
+     *   entry for IsLatest versions
      */
-    _getBucketObjects(bucketList, done) {
-        if (!bucketList) {
-            return done(null, null);
+    _getObjectVersionsList(bucket, state, done) {
+        if (bucket === constants.usersBucket ||
+            bucket === constants.metastoreBucket) {
+            return done();
         }
-        return async.mapLimit(bucketList, 10, (bucketInfo, cb) => {
-            if (bucketInfo === constants.usersBucket ||
-            bucketInfo === constants.metastoreBucket) {
-                return cb();
+        const { versionMarker, keyMarker } = state;
+        const params = {
+            Bucket: bucket,
+        };
+        // if previous state, should paginate here
+        if (versionMarker && keyMarker) {
+            params.VersionIdMarker = versionMarker;
+            params.KeyMarker = keyMarker;
+        }
+        // TODO: For testing, I can set MaxKeys
+        const req = this._s3Client.listObjectVersions(params);
+        attachReqUids(req, this.requestLogger);
+        return req.send((err, data) => {
+            if (err) {
+                this.log.error('error getting list of object versions', {
+                    method: 'IngestionProducer._getObjectVersionsList',
+                    error: err,
+                    bucket,
+                });
+                return done(err);
             }
-            const req = this.ringReader.getObjectList({
-                Bucket: bucketInfo,
-            });
-            attachReqUids(req, this.requestLogger);
-            return req.send((err, data) => {
-                if (err) {
-                    this.log.error('error getting list of objects', {
-                        method: 'IngestionProducer:getBucketObjects', err });
-                    return cb(err);
-                }
-                return cb(null, { bucket: bucketInfo, objects: data.Contents });
-            });
-        }, (err, buckets) => done(err, buckets));
+            const {
+                IsTruncated,
+                NextKeyMarker,
+                NextVersionIdMarker,
+                Versions,
+                DeleteMarkers
+            } = data;
+
+            const versionList = [...Versions, ...DeleteMarkers];
+            // find all IsLatest versions since we need these as separate
+            // metadata entries
+            // const latestVersions = versionList.filter(v => v.IsLatest)
+            //     .map(v => Object.assign({}, v, { isLatest: true }));
+
+            const response = {
+                versionList,
+                IsTruncated,
+            };
+            if (IsTruncated) {
+                response.versionMarker = NextVersionIdMarker;
```

Comment:
Minor: these can just be added to the `response` object since they'll be `undefined` anyway if it isn't truncated, and you check for that later.